[Feature] 배틀 입장 이벤트 구현

HoonDongKang opened the pull request:
## 🔗 관련 이슈

<!-- 예: Closes #123, Relates to #456 -->

Closes #8

## ✅ 작업 내용

### 💡개선 사항

<!-- 이 PR에서 변경된 주요 내용 개선 사항 -->

1.  `battle:join` 이벤트 구현
    1. 소켓 연결
       1. 유저 권한 확인 & 유저 메타 데이터 등록
    2. 배틀 참여
       1. 유효한 `battleId`로 접근하였는가
       2. 만약 password가 있다면 비밀번호가 일치하는가
       3. 이미 종료된 배틀인가
    3. 해당 배틀 ID의 진행 상태를 조회(2번에서 설명)
    4. 해당 참여자 `battle:${battleId}` Room으로 참여
    5. `battle:joined` 이벤트 발행으로 배틀 현황을 브로드캐스팅

2.  `battle:joined` 이벤트 발행 - 사용자가 배틀을 참여하면 해당 배틀에 대한 데이터를 응답한다.
    1.  `battleId`로 Redis에서 배틀에 관한 데이터 조회
        1.  투표 수
        2.  이의제기 리스트
        3.  반박 리스트
        4.  해당 진영 채팅
    2.  나 + 브로드캐스팅으로 전달

## 📸 참고 사항

<!-- 코드 리뷰에서 참고할 사항, 새로 주의 깊게 봐야 하는 파일/코드, 주요 고민과 해결 과정-->
![JOIN](https://github.com/user-attachments/assets/98355387-bc99-4913-bcf9-9e4f1cf69d2f)

<br />

## 💬 질문사항 (고민했던 부분)

<!-- 리뷰어에게 묻고 싶은 것, 트레이드오프 등 -->

1. 하나의 Battle을 Room으로 봐야할 지, Battle 내 각 진영을 Room으로 할 지 고민입니다.
   - Room이 많으면 역할이 명확하게 구분되지만 진영을 옮길 때마다 disconnect와 connect가 빈번하게 이루어지며 통신에 대한 지연이 발생할 수도 있다고 판단하였습니다.
   - 그리고 Room 이동 과정에서 메세지를 잃을 수도 있다고 생각하였습니다
   - 일단 프로토타입은 Battle 별로 Room 생성
2. Redis가 일단 연결되면 그 때 Redis 내에서 배틀 연결을 실질적으로 해볼 수 있을 것 같습니다..!!

HoonDongKang committed:
feat: Socket.io 의존성 추가

HoonDongKang committed:
feat: Socket.io 의존성 추가

HoonDongKang committed:
feat: Battle Module 추가

HoonDongKang committed:
feat: Battle Module 추가

HoonDongKang committed:
feat: 배틀 참여 DTO 구현

HoonDongKang committed:
feat: 배틀 참여 DTO 구현

HoonDongKang committed:
feat: 소켓 연결 및 배틀 참여

HoonDongKang committed:
feat: 소켓 연결 및 배틀 참여

HoonDongKang committed:
feat: 사용자 배틀 참여 및 배틀 동기화

HoonDongKang committed:
feat: 사용자 배틀 참여 및 배틀 동기화

HoonDongKang committed:
Merge branch 'dev' of https://github.com/boostcampwm2025/web02-CMC into feature/#28-battle-join

HoonDongKang committed:
fix: implement onGatewayConnection, Disconnect

jjjjjinseo commented on backend/src/battles/battles.service.ts:
```diff
@@ -0,0 +1,126 @@
+    // async joinBattle(dto: JoinBattleDto, socketId: string, userId: string) {
+    const { battleId, password } = joinBattleDto
+
+    if (!battleId) {
```
예외처리까지 꼼꼼히 해주셨네요 💯
나중에 에러 핸들링 적용해도 좋을 것 같아요!

jjjjjinseo reviewed (approved):
수고하셨습니다! 소켓 마스터가 되셨네요 🙂

고민하신 부분에 대해서는, 제 생각에도 진영 변경 기능 때문에 현재처럼 배틀과 room을 1:1로 취급하는 구조가 적절해 보입니다!

rhxfkg reviewed (approved):
수고하셨습니다!
고민하신 내용에 대해 생각해봤는데 메시지 유실되는 상황이 최악인 것 같아 배틀 = 룸으로 진행하는 것이 좋다고 생각합니다.

U-S-jun commented on backend/src/battles/battles.gateway.ts:
```diff
@@ -0,0 +1,60 @@
+
+      client.emit('battle:joined', battleState)
+
+      client.to(`battle:${battleId}`).emit('battle:joined', { ...battleState, participantId: client.id })
```
현재 페이로드는 클라이언트가 battleState필드로 중첩되어 전달받을 것 같은데 구조 분해 후 emit을 하는건 어떨까요?

U-S-jun commented on backend/src/battles/battles.gateway.ts:
```diff
@@ -0,0 +1,60 @@
+    } catch {
+      client.disconnect()
+    }
+  }
```
소켓 연결이 되면 진영 선택으로 넘어가도록 해야하니 이후 여기서 각 팀 코드를 보내주도록 작성되면 될까요?

U-S-jun commented on backend/src/battles/dto/join-battle.dto.ts:
```diff
@@ -0,0 +1,5 @@
+  battleId: string
+  team: string
+  password?: string
+}
```
해당 DTO 에서는 validator 사용하지 않아도 될까요??

U-S-jun reviewed (commented):
수고하셨습니다!
API 명세서에서 POST join은 필요가 없을 것 같네요 이따 스크럼 때 논의하고 수정하도록 할게요!

U-S-jun committed:
[Feature]: battle create api (#9)

* feature(be): 배틀 생성 라우팅 및 요청 body 검증

* feature(be): 타입 생성

* feature(be): 배틀 생성 서비스 구현

* test(be): 배틀 생성 단위테스트 작성

* feature(be): 모듈 추가

* feature(be): 리뷰 반영

상수 관리 : BATTLE_STATUS
validation pipe 도입
isPublic 삭제
UUIDv7 도입
DTO 폴더 관리
폴더 구조 컨벤션 반영

* feature(be): validation pipe

* 의존성 추가

* 의존성 추가

jjjjjinseo committed:
[Feature] 배틀 목록 조회 API 구현 (#4)

* feature: battles 도메인 기본 enum/const 타입 정의 #4

* fix: lint 설정 중 cd 제거 및 eslint 설정 파일 주입 #4

* feat: mock battle data 추가 #4

* feat: 실시간 배틀 목록 조회 기능 추가 #4

* feat: 지난 배틀 조회 API 구현 #4

* fix: 목록 조회 시 최신순/최신 종료 순으로 정렬 #4

* fix: battle 필드 정리 #4

* test: battle service 중 목록 조회 단위테스트 #4

* test: battle 컨트롤러, dto 유효성 테스트 #4

* fix: class 명 복수형으로 수정 #4

* fix: enum 형태 as const, type으로 변경 #4

* fix: status open,closed로 통합 #4

* chore: battle query dto 파일 네이밍 변경 #4

* docs: todo 주석 추가 #4

* resolve conflict

rhxfkg committed:
[Feature] 배틀 결과 조회 API 구현하기 (#13)

* test: 배틀 결과 조회 테스트 코드 작성 #13

* feat: 배틀 결과 목데이터 생성 #13

* feat: 배틀 결과 DTO 정의 #13

* feat: 배틀 결과 조회 API 구현 #13

* refactor: 목데이터 사용으로 불필요한 async/await 제거 #13

* chore: CORS 설정 추가 #13

* style: lint-staged 포맷팅 적용 #13

* refactor: fromEntity 패턴 적용 및 타입 안정성 개선 #13

- BattleResultResponseDto에 fromEntity 정적 메서드 추가
- Service에서 스프레드 연산자 대신 fromEntity 사용
- calculateMVP에서 타입 단언 제거 (Map 타입 명시)

* refactor: HttpException 통합 및 CORS 환경변수화 #13

- Controller에서 HttpException으로 에러 처리 통합
- CORS origin을 환경변수로 관리 (FRONTEND_URL)
- .env.example 파일 추가 (환경변수 템플릿)

* refactor: DTO와 Type 분리 #13

- types/battleResult.types.ts 생성하여 도메인 타입 정의
- DTO는 API 응답용 class만 유지
- Service에서 타입 import 경로 수정
- 관심사 분리 및 재사용성 향상

* refactor: 폴더 구조 재구성 #13

- controller/, service/ 폴더로 분리
- mock 파일을 battles/mock/으로 이동
- import 경로 수정
- dev 브랜치 구조에 맞춰 정리

* refactor: dev 브랜치와 병합 준비 #13

- battles.service.ts: dev와 우리 브랜치 메서드 통합
- battles.controller.ts: 4개 엔드포인트 통합 (라우트 순서 조정)
- battles.controller.spec.ts: mock 기반 테스트로 통일
- battles.service.spec.ts: dev 방식으로 통일
- uuid 패키지 설치
- 중복 파일 제거 (battles.controller.spec.ts)
- 모든 테스트 통과 (22/22)

---------

Signed-off-by: rhxfkg <51471056+rhxfkg@users.noreply.github.com>

HoonDongKang committed:
fix: 의존성 설치 conflict

node_modules, pnpm-lock.yaml 삭제 후 재설치

HoonDongKang committed:
chore: 중복된 타입 제거 및 상수 처리

- const 파일 내부의 type 들을 types 폴더로 이동
- 중복되는 type들 통합
- 도메인 로직에서 매직넘버 상수처리

HoonDongKang committed:
chore: 폴더 구조 정리

HoonDongKang committed:
feat: 현재 진행 중인 배틀 타입 생성

HoonDongKang committed:
feat: 사용자 배틀 참여 소켓 연결 기능 구현

1. ActiveBattleStates 라는 Map 구조를 통해 현재 진행 중인
배틀 상태를 저장합니다.
2. 배틀을 생성할 떄, BattleState이 초기화됩니다.
3. 사용자가 참여하면 해당 BattleState에 유저 정보를 추가합니다.
4. battle:joined에 대한 응답으로 현재 유저 정보가 추가된 배틀 상태를 반환

HoonDongKang committed:
style: 파일명 CamelCase로 변경

HoonDongKang committed:
feat: POST /battles/:id/join  API 구현

HoonDongKang committed:
test: Battle Join 테스트 코드 추가

HoonDongKang committed:
Merge branch 'feature/#28-battle-join' of https://github.com/boostcampwm2025/web02-CMC into feature/#28-battle-join

HoonDongKang commented:
1. 전체적으로 중복되는 타입이나 상수들 정리하여 도메인 로직 속 매직 넘버들을 상수처리하였습니다.
2. 현재 `Battles`로 배틀 데이터를 저장하고 있는데, 배틀 진행에 필요한 데이터를 함께 저장하면 데이터가 불필요하게 커질 것 같다고 판단하여 `activeBattles`라는 Map 자료구조를 추가하여 관리하였습니다.
     -  배틀 생성 및 종료는 DB 저장(`Battles`)
     -  배틀 진행 상황 저장은 Redis 저장(`ActiveBattles`)
     - 이러한 방식으로 분리하여 자료구조를 하나 추가하였습니다.
3. 배틀 생성 시, `Battles`에 배틀에 관한 속성이 추가되고 해당 배틀 진행을 위한 State를 `activeBattles`에 초기화시켜주었습니다.
4. 참여자가 진영 페이지로 가면 `POST /battes/:id/join`으로 배틀에 관한 데이터를 받아옵니다. 
    - `title, descrption, aCode, bCode`
5. 진영을 선택하면 해당 배틀에 소켓이 연결됩니다.
    - `all`, `teamA`, `teamB`로 Room 변수명 구분
6. 배틀 진행에 필요한 interface는 `ActiveBattleState` 로 정의하였습니다.

HoonDongKang commented:
![join](https://github.com/user-attachments/assets/97f456d1-1d38-4d4b-aa27-e5424ef1ee70)

U-S-jun reviewed (approved):
수고하셨습니다!
